ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp

```diff
@@ -33,14 +33,56 @@ StackRenderer::render_thread_begin(PyThreadState* tstate,
     // clock_gettime(CLOCK_MONOTONIC) on linux and mach_absolute_time() on macOS.
     // This is not the same as std::chrono::steady_clock, which is backed by clock_gettime(CLOCK_MONOTONIC_RAW)
     // (although this is underspecified in the standard)
+    int64_t now_ns = 0;
     timespec ts;
     if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
-        auto now_ns = static_cast<int64_t>(ts.tv_sec) * 1'000'000'000LL + static_cast<int64_t>(ts.tv_nsec);
+        now_ns = static_cast<int64_t>(ts.tv_sec) * 1'000'000'000LL + static_cast<int64_t>(ts.tv_nsec);
         ddup_push_monotonic_ns(sample, now_ns);
     }
 
+    // Save the thread information in case we observe a task on the thread
+    thread_state.id = thread_id;
+    thread_state.native_id = native_id;
+    thread_state.name = std::string(name);
+    thread_state.now_time_ns = now_ns;
+    thread_state.wall_time_ns = 1000LL * wall_time_us;
+    thread_state.cpu_time_ns = 0; // Walltime samples are guaranteed, but CPU times are not. Initialize to 0
+                                  // since we don't know if we'll get a CPU time here.
+
+    // Finalize the thread information we have
     ddup_push_threadinfo(sample, static_cast<int64_t>(thread_id), static_cast<int64_t>(native_id), name);
-    ddup_push_walltime(sample, 1000LL * wall_time_us, 1);
+    ddup_push_walltime(sample, thread_state.wall_time_ns, 1);
+}
+
+void
+StackRenderer::render_task_begin(std::string_view name)
+{
+    static bool failed = false;
+    if (failed) {
+        return;
+    }
+    if (sample == nullptr) {
+        // The very first task on a thread will already have a sample, since there's no way to deduce whether
+        // a thread has tasks without checking, and checking before populating the sample would make the state
+        // management very complicated.  The rest of the tasks will not have samples and will hit this code path.
+        sample = ddup_start_sample();
+        if (sample == nullptr) {
+            std::cerr << "Failed to create a sample.  Stack v2 sampler will be disabled." << std::endl;
+            failed = true;
+            return;
+        }
+
+        // Add the thread context into the sample
+        ddup_push_threadinfo(sample,
+                             static_cast<int64_t>(thread_state.id),
+                             static_cast<int64_t>(thread_state.native_id),
+                             thread_state.name);
+        ddup_push_walltime(sample, thread_state.wall_time_ns, 1);
+        ddup_push_cputime(sample, thread_state.cpu_time_ns, 1); // initialized to 0, so possibly a no-op
```

Comment:
`thread_state.cpu_time_ns` will will be zero only when `render_cpu_time` has not yet been called, right?